src/navigation/path_planner.py

```diff
@@ -0,0 +1,242 @@
+            completed, such that get_intermediate_target() returns the next target in the
+            path the next time it is called.
+        - is_path_complete(): returns True/False depending on whether the overall target
+            has been marked as completed.
+        - generate_path(Point src, Point target): This should not typically be called.
+            This overrides the caching protocol and forcefully generates a new path, even
+            if the target_pos has not changed. The path can then be accessed normally,
+            using get_intermediate_target() and complete_intermediate_target()
+    """
+
+    failure_zones: List[FailureZone]
+    visibility_graph: Graph
+
+    source_pos: Point
+    target_pos: Point
+    path: List[Point]
+    target_vertex_idx: int
+
+    def __init__(self):
+        self.failure_zones: List[FailureZone] = []
+        self.visibility_graph: Graph = Graph(undirected=True)
+
+        self.source_pos: Point = None
+        self.target_pos: Point = None
+        self.path: List[Point] = None
+        self.target_vertex_idx: int = 0
+
+    def get_intermediate_target(self, source_pos: Point, target_pos: Point) -> Point:
+        """
+         Given a source position (the current position of the rover) and a target position
+         (the final destination), generates an intermediate target position after
+         calculating a clear path between source and destination.
+
+         This path is recalculated only when the target position changes, or a FailureZone
+         is added to the graph. Once the rover reaches this intermediate target,
+         complete_intermediate_target() should be called so that the path can update and
+         return the next target on the next call to get_intermediate_target().
+
+         Note: As the path is only recalculated when the target position changes, it is
+               assumed that the source_pos of the rover remains along this path. If you
+               you want to override this functionality and generate a path, call
+               generate_path() before calling this funciton.
+
+        :param source_pos:   Point object representing source position
+        :param target_pos:   Point object representing target position
+
+        :return:             A Point object representing the intermediate target
+        """
+        if (not self.path) or (not target_pos == self.target_pos):
+            self.generate_path(source_pos, target_pos)
+
+        if self.target_vertex_idx == len(self.path):
+            return self.path[-1]
+
+        return self.path[self.target_vertex_idx]
+
+    def complete_intermediate_target(self) -> None:
+        """
+        Marks the previously-returned intermediate target as complete.
+        """
+        assert self.path
+
+        if self.target_vertex_idx < len(self.path):
+            self.target_vertex_idx += 1
+
+    def is_path_complete(self) -> bool:
```

Comment:
I'm not so sure about that because the Path is pretty directly integrated with the PathPlanner, so it would be a lot clunkier to have it in a separate class. The is_path_complete is used by the Driver, so it has to exist in the PathPlanner. Basically, the PathPlanner interface works mainly thru the methods of adding failure zones, getting the next vertex, marking vertices as completed, and checking whether the path has been completed (is_path_complete) in terms of all vertices being marked as completed.